### PR TITLE
WIP - Experiment replacing the vendor dir with a submodule

### PR DESCRIPTION
**Pull request by smarterclayton**

### CODE REVIEW

**Comment by deads2k** [comment]
@pmorie You had strong feeling submodules.  Can you speak specifically to what this does?

File: .gitmodules [hunk 1]
```diff
@@ -0,0 +1,5 @@
+[submodule "kubernetes"]
+	path = vendor/k8s.io/kubernetes
+	url = https://github.com/openshift/kubernetes.git
+  branch = openshift-3.7
```
Comment (deads2k) [inline .gitmodules]: release-1.7.0

**Comment by smarterclayton** [comment]
See the latest makefile. `make vendor` is a target that ensures that the vendor directory in kube is in the "sparse" state (non-existent).  `make vendor/k8s.io/kubernetes/vendor` ensures the vendor directory is in the state of being checked out / not-sparse.

So in theory `make ANYTHING_ORIGIN` can depend on the `vendor` task - and the vendor tree matches origin.  `make ANYTHING_KUBE_SPECIFIC` can depend on the `vendor/k8s.io/kubernetes/vendor` task and be in the right state (i.e. you can depend on that and then run make / test / build inside of the kube directory).

**Comment by sttts** [comment]
The checkout part of the process looks feasible. How would one do kube changes?

If I change the sparse checkout of `vendor/k8s.io/kubernetes`, I have to push my changes somewhere, probably to my private fork `github.com/sttts/kubernetes`. Then I push origin to my fork. 

The problem: nobody can see my kube commit as it is in a different repo (also CI can't!). I.e. you have to change .gitmodules to point to my fork. On merge we would have to merge the kube PR, revert the .gitmodules change and then merge the origin PR.

**Comment by detiber** [comment]
> The checkout part of the process looks feasible. How would one do kube changes?

Wouldn't they just be done against the k8s fork? 

> If I change the sparse checkout of vendor/k8s.io/kubernetes, I have to push my changes somewhere, probably to my private fork github.com/sttts/kubernetes. Then I push origin to my fork.

For local testing/hacking you could point it to a local file url, which would avoid the push step. Ideally, though I think one should be able to test local changes to the submodule in tree without hacking around the use of submodules.

> The problem: nobody can see my kube commit as it is in a different repo (also CI can't!). I.e. you have to change .gitmodules to point to my fork. On merge we would have to merge the kube PR, revert the .gitmodules change and then merge the origin PR.

I would expect CI to handle updating the submodule references rather than having manual steps. Maybe something like this:
k8s fork PR -> k8s fork CI pipeline -> merge PR -> run origin pipeline w/ updated submodule -> merge submodule update through submit queue w/ high priority.

Of course that wouldn't work for more intertwined changes or rebases, but those could have customized

File: .gitmodules (hunk 1 above)
Comment (detiber) [reply .gitmodules]: Not saying you are wrong, but openshift-specific branch naming might help reduce confusion for new contributors.

**Comment by smarterclayton** [comment]
The makefile has the targets that flip the checkout mode.  make vendor
and origin builds.  make vendor/k8s.io/kubernetes/vendor and you can
run Kube build commands in the vendor dir.

**Comment by sttts** [comment]
> k8s fork PR -> k8s fork CI pipeline -> merge PR -> run origin pipeline w/ updated submodule -> merge submodule update through submit queue w/ high priority.

Was talking more about the non-merge case where we want to integration test our two changes. Not saying it's undoable, but the CI has understand that there is another PR in another repo (kube fork) with the same HEAD as the submodule. It could derive the Github repo from that (i.e. your private fork) and check it out (by possible changing the .submodules file to point to your fork). All doable, but one of the messy details.

**Comment by sttts** [comment]
> The makefile has the targets that flip the checkout mode.  make vendor
> and origin builds.  make vendor/k8s.io/kubernetes/vendor and you can
> run Kube build commands in the vendor dir.

Sounds all nice on paper. I fear we miss a couple of messy details if we start to model the complete process including CI.

**Comment by sttts** [comment]
Related, golang/dep's position around submodules: https://github.com/golang/dep/issues/1240

**Comment by smarterclayton** [comment]
What about dep actually being useful?  Do they have a plan for that yet?

Jokes aside, this flow benefits joint code changes in origin and Kube
simultaneously at the expense of ci complexity.  If we think we can create
a sustainable model without direct "commit to Kube and origin together"
then we don't need this.


On Oct 13, 2017, at 3:07 PM, Dr. Stefan Schimanski <notifications@github.com>
wrote:

Related, golang/dep's position around submodules: golang/dep#1240
<https://github.com/golang/dep/issues/1240>

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/16685#issuecomment-336541564>,
or mute
the thread

**Comment by smarterclayton** [comment]
I'll note that:


   - Nesting one repository within another implies that changes could,
   potentially, be made directly in that subrepository. This is directly
   contrary to dep's foundational principle that vendor is dead code, and
   directly modifying anything in there is an error.


Is in direct contradiction to our entire development model.

**Pull request closed without merging**